### PR TITLE
Insight metrics

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Metrics](https://tools.hmcts.net/jira/browse/RPE-210)

### Change description ###

Track events of message receive and acknowledgement

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.reform.sendletter.logging;
+
+public final class AppEvent {
+
+    /**
+     * Used to track event prior converting letter to servicebus message to track all income.
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java]: income?

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java]: perhaps i meant traffic 😂  or smth like that

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 2]
```diff
@@ -86,24 +97,26 @@ public void should_throw_connection_exception_when_queue_client_fails_to_connect
...
         verify(queueClientSupplier).get();
-        verifyNoMoreInteractions(queueClientSupplier);
+        verifyNoMoreInteractions(queueClientSupplier, queueClient, insights);
     }
 
     @Test
     public void should_throw_json_processing_exception_letter_serialization_fails() throws Exception {
         // given
-        given(objectMapper.writeValueAsBytes(any())).willThrow(JsonProcessingException.class);
+        willThrow(JsonProcessingException.class).given(objectMapper).writeValueAsBytes(any());
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: is this related to insights in any way?

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 3]
```diff
@@ -55,6 +62,7 @@ public void should_return_message_id_when_single_letter_is_sent() throws Excepti
         // given
         given(queueClientSupplier.get()).willReturn(queueClient);
         given(queueClient.sendAsync(any(Message.class))).willReturn(voidCompletableFuture);
+        // given(queueClient.closeAsync()).willReturn(voidCompletableFuture);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: ?

File: src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.reform.sendletter.logging;
+
+public final class AppEvent {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java]: Is there a reason why the two constants can't be in the class itself where it is used?

Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java]: As insight class may grow with additional functionality in there, event with descriptions too - good point to keep utility class separate

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java]: I think we should never code for requirements which are unknown.In future something changes you can create a utility class if required.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppEvent.java]: this is like an enum, imho good to group it somehow

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java [hunk 5]
```diff
@@ -49,10 +59,39 @@ public void should_fail_creating_app_insights_when_instrumentation_key_is_not_pr
...
+
+    @Test
+    public void should_track_message_received_event() {
+        context.setInstrumentationKey(IKEY);
+
+        Map<String, String> properties = new HashMap<>();
+
+        properties.put("service", SERVICE_NAME);
+        properties.put("template", TEMPLATE);
+        properties.put("messageId", MESSAGE_ID);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: you can use Guava's immutable map for nicer initialization

Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: `properties` doesn't say much, and the fact that this map is created in setup stage of the test doesn't help.
Perhaps it would be clearer if this was moved to verification part of the test? Even inlined if immutable map used

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 6]
```diff
@@ -64,7 +71,14 @@ public void should_return_message_id_when_single_letter_is_sent() throws Excepti
 
         verify(queueClientSupplier).get();
         verify(queueClient).sendAsync(any(Message.class));
-        verifyNoMoreInteractions(queueClientSupplier, queueClient);
+
+        voidCompletableFuture.thenRun(() -> {
+            verify(queueClient).closeAsync();
+            verify(insights).trackMessageAcknowledgement(messageId);
+            verify(insights).trackMessageReceived("service", letter.template, messageId);
+            verifyNoMoreInteractions(queueClientSupplier, queueClient);
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: Do we need both of these ?

Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: if you check for interaction of mocks you have in one test case scope - best to include all of them. in fact, i'm missing `insights` there. probably after investigation as to why the tests did not work on travis

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 7]
```diff
@@ -47,23 +54,42 @@ public String send(Letter letter, String serviceName) throws JsonProcessingExcep
...
+        Message message = createQueueMessage(new WithServiceName<>(letter, serviceName), messageId);
+        Instant startSending = Instant.now();
+
+        CompletableFuture<Void> sendResult = sendClient
+            .sendAsync(message)
+            .whenCompleteAsync((result, exc) -> {
+                Duration tookSending = Duration.between(startSending, Instant.now());
+                insights.trackMessageAcknowledgement(tookSending, exc != null, messageId);
+
+                if (exc != null) {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: you could extract the null check to a variable as it is getting repeated `boolean isException=exc != null;`

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: Also you could refactor the code as below
```
 CompletableFuture<Void> sendResult = sendClient
            .sendAsync(message)
            .whenCompleteAsync(onCompleteAsync(messageId, startSending))
            .thenRunAsync(sendClient::closeAsync);
```

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 8]
```diff
@@ -47,23 +54,42 @@ public String send(Letter letter, String serviceName) throws JsonProcessingExcep
...
+            throw new SendMessageException("Could not send message to ServiceBus with " + messageId);
+        }
 
         return messageId;
     }
 
     private Message createQueueMessage(WithServiceName<Letter> letter, String messageId)
         throws JsonProcessingException {
 
+        insights.trackMessageReceived(letter.service, letter.obj.template, messageId);
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: I think message is received earlier than this point.At this point we are just converting it to a form which Azure API can understand.

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java [hunk 9]
```diff
@@ -49,10 +61,77 @@ public void should_fail_creating_app_insights_when_instrumentation_key_is_not_pr
...
 
         assertThat(TelemetryConfiguration.getActive().getChannel().isDeveloperMode()).isFalse();
     }
+
+    @Test
+    public void should_track_message_acknowledgement_event_for_success_case() {
+        context.setInstrumentationKey(IKEY);
+
+        AppInsights insights = new AppInsights(telemetry);
+        Map<String, String> properties = Collections.singletonMap("messageId", MESSAGE_ID);
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: we can use `ImmutableMap.of("messageId", MESSAGE_ID);` here

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java [hunk 10]
```diff
@@ -49,10 +61,77 @@ public void should_fail_creating_app_insights_when_instrumentation_key_is_not_pr
...
+        verify(telemetry).trackDependency(
+            AppInsights.SERVICE_BUS_DEPENDENCY,
+            AppInsights.SERVICE_BUS_MESSAGE_ACKNOWLEDGED,
+            new com.microsoft.applicationinsights.telemetry.Duration(timeTook.toMillis()),
+            false
+        );
+    }
+
+    @Test
+    public void should_track_exception() {
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: what it is scenario you are testing here.test method name doesn't say much

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java [hunk 11]
```diff
@@ -49,10 +61,77 @@ public void should_fail_creating_app_insights_when_instrumentation_key_is_not_pr
...
+        insights.trackMessageAcknowledgement(timeTook, true, MESSAGE_ID);
+
+        verify(telemetry).trackDependency(
+            AppInsights.SERVICE_BUS_DEPENDENCY,
+            AppInsights.SERVICE_BUS_MESSAGE_ACKNOWLEDGED,
+            new com.microsoft.applicationinsights.telemetry.Duration(timeTook.toMillis()),
+            true
+        );
+        verify(telemetry).trackEvent(AppInsights.SERVICE_BUS_MESSAGE_ACKNOWLEDGED, properties, null);
+    }
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: Are you sure there are no more interactions ?It would be good if we add `verifyNoMoreInteractions(telemetry);`

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java [hunk 12]
```diff
@@ -49,10 +61,77 @@ public void should_fail_creating_app_insights_when_instrumentation_key_is_not_pr
...
+
+        insights.trackMessageAcknowledgement(timeTook, false, MESSAGE_ID);
+
+        verify(telemetry).trackDependency(
+            AppInsights.SERVICE_BUS_DEPENDENCY,
+            AppInsights.SERVICE_BUS_MESSAGE_ACKNOWLEDGED,
+            new com.microsoft.applicationinsights.telemetry.Duration(timeTook.toMillis()),
+            false
+        );
+    }
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: It would be good if we add verifyNoMoreInteractions(telemetry);

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 13]
```diff
@@ -98,12 +146,13 @@ public void should_throw_json_processing_exception_letter_serialization_fails()
         Throwable exception = catchThrowable(() -> service.send(letter, "service"));
 
         // then
-        assertThat(exception)
-            .isInstanceOf(JsonProcessingException.class);
+        assertThat(exception).isInstanceOf(JsonProcessingException.class);
+        verify(insights, never()).trackMessageAcknowledgement(any(Duration.class), anyBoolean(), anyString());
+        verify(insights).trackMessageReceived(eq("service"), eq(letter.template), anyString());
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: `verifyNoMoreInteractions` missing

File: src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppInsights.java [hunk 14]
```diff
@@ -1,13 +1,56 @@
...
+
+    public void trackMessageReceived(String serviceName, String template, String messageId) {
+        Map<String, String> properties = ImmutableMap.of(
+            "service",
+            serviceName,
+            "template",
+            template,
+            "messageId",
+            messageId
+        );
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppInsights.java]: that would be way easier to read imho:
```
       Map<String, String> properties = ImmutableMap.of(
           "service", serviceName,
           "template", template,
           "messageId", messageId
       );
```
I hope checkstyle will allow this

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java (hunk 7 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: no it cannot be refactored as we specifically want to capture success/failure of message send event, which is the first call to the client

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java (hunk 9 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: @nitinprabhu `singletonMap` returns an immutable map as well, no need

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java [hunk 15]
```diff
@@ -49,10 +61,77 @@ public void should_fail_creating_app_insights_when_instrumentation_key_is_not_pr
...
+        insights.trackMessageReceived(SERVICE_NAME, TEMPLATE, MESSAGE_ID);
+
+        Map<String, String> properties = ImmutableMap.of(
+            "service",
+            SERVICE_NAME,
+            "template",
+            TEMPLATE,
+            "messageId",
+            MESSAGE_ID
+        );
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: same as in comment above about readability

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java (hunk 9 above)
Comment (nitinprabhu) [reply src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: @rkondratowicz Reason I thought is because at most of the places Immutable map is used it's only at this place we are using singleton.

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java (hunk 8 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: this is logging the incoming request to print some sort of letter. since serialisation can throw exception and we might loose the information of the message we want to track - the insights is done right before that. especially when it is not needed to track any time related metric

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java (hunk 9 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: don't see the point of using different library's implementation when it is available in java itself. just a singleton map 😃

File: src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java (hunk 10 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/sendletter/logging/AppInsightsTest.java]: would it be more preferable if the title said `should track any exception`? really the point is to just test out functionality works

File: src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppInsights.java (hunk 14 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/logging/AppInsights.java]: didn't risk initially to not pester checkstyle :D
